qiskit/passmanager/flow_controllers.py

```diff
@@ -10,149 +10,201 @@
 # copyright notice, and modified files need to carry a notice indicating
 # that they have been altered from the originals.
 
-"""Pass flow controllers to provide pass iterator conditioned on the property set."""
+"""Built-in pass flow controllers."""
 from __future__ import annotations
-from collections import OrderedDict
-from collections.abc import Sequence
-from typing import Union, List
-import logging
 
-from .base_pass import GenericPass
+import logging
+from collections.abc import Iterator, Callable
+from typing import Type
+
+from .base_optimization_tasks import (
+    BaseFlowController,
+    OptimizerTask,
+)
+from .propertyset import PropertySet
 from .exceptions import PassManagerError
 
 logger = logging.getLogger(__name__)
 
 
-class FlowController:
-    """Base class for multiple types of working list.
+class FlowControllerMeta(type):
+    """A magic to separate controller factory and flow control.
+
+    FlowController provides a factory function to build a nested flow controller,
+    but this class is also designed as a baseclass of all flow controllers.
+
+    This structure causes unmanageable mix-up of factory function and flow control,
+    yielding something awkward such that a user may instantiate a flow controller
+    with FlowController.controller_factory to get its instance,
+    and one can further build another controller instance from it.
+
+    When a user subclasses the FlowController, there is no guarantee that
+    the user always adds their controller to the base FlowController namespace.
+    When a controller is added to the subclass, the base FlowController cannot
+    look up the alias in its own namespace.
+
+    This metaclass splits the responsibility of flow control from the FlowController
+    and delegates it to the BaseFlowController, while allowing users to
+    directly subclass the FlowController for "backward compatibility".
```

Comment:
What's our path forwards for getting rid of this whole structure?  What would the flow-controller API and class structure look like if we didn't have to worry about the backwards compatibility at all?

I'm just wondering if there's more we could do to move into a sensible position more directly: I think in practice, it's probably not really actually possible to subclass `FlowController` in current Qiskit and have it really work, so _maybe_ there's some scope for making larger changes than we usually would.

Reply:
This is more about API of pass manager. Currently pass manager `.append` supports keyword arguments that implicitly instantiate flow controllers with its alias, e.g.
https://github.com/Qiskit/qiskit/blob/a745993c585aa062a520eb4cb0bfd82742999896/qiskit/transpiler/preset_passmanagers/builtin_plugins.py#L628

I don't think this drastically improves the readability. When multiple aliases are set in the kwargs, priority of flow controller is implicitly defined by the class hierarchy defined in the `FlowController`. I prefer defining the controller more explicitly before adding them to the pass manager. For example,

https://github.com/Qiskit/qiskit/blob/5034cc49d2767ec1ef13ac526609af38d15119bb/qiskit/passmanager/__init__.py#L140-L159

Reply:
e.g. https://github.com/Qiskit/qiskit/blob/a745993c585aa062a520eb4cb0bfd82742999896/test/python/transpiler/test_pass_scheduler.py#L183-L190

Reply:
Thanks, I also wanted to drop this logic if possible. I removed the metaclass in [59837b3](https://github.com/Qiskit/qiskit/pull/10127/commits/59837b35ab2f7211ffafb98ffe8e2bad61b330ca), which may break some user code with typecheck because `FlowControllerLinear`, `DoWhileController`, and `ConditionalController` are not subclass of `FlowController`. I needed to fix one unittest as well. I assume impact of this change is not significant. Then I deprecated factory method in [ca7dba6](https://github.com/Qiskit/qiskit/pull/10127/commits/ca7dba6bf215763f1f923fefa3cdc9d8e459e3bc). The `_legacy_build_flow_controller` still needs to depend on the `FlowController` class because user (and some unittests) can define custom controller subclass and register it to the class variable to use the factory method. Eventually we should remove `FlowController` along with the kwargs syntax.